[Tan Yu Fei] iP

Lanreath opened the pull request:
# DukePro
> "Duke is the greatest piece of software ever written." - Lanreath
DukePro frees your mind of having to remember things. Just do it.
- text-based
- easy to learn
- ~~FAST~~ _REAL_ FAST
All you need to do is:
1. download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html)
2. double-click it.
3. add your tasks.
4. let it remember tasks for you :grimacing:
And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
---
If you a Java lover, you can use it practice Java too. (_but why would you?_) Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Lanreath committed:
Echo user console input, exit on bye

Lanreath committed:
Store text and display when requested

Lanreath committed:
Task marking/unmarking

Lanreath committed:
Add support for three task types, refactored tasks into package, todo refactoring on conditionals

Lanreath committed:
Added testcases and modified test script

Lanreath committed:
Added user input validation with DukeException, refactored commands to methods

Lanreath committed:
Added deletion

Lanreath committed:
Add automatic task storage on hard disk

Lanreath committed:
Add datetime parsing and output

Lanreath committed:
Merge branch 'branch-Level-7'

Lanreath committed:
Merge branch 'branch-Level-8'

Lanreath committed:
Fix parsing bug

Lanreath committed:
Refactor reused objects into classes of Parser, Storage, TaskList, Ui, Command etc.

Lanreath committed:
Organised classes into packages

Lanreath committed:
Merge branch 'add-gradle-support'

Lanreath committed:
Use Gradle to automate build and run

Lanreath committed:
Add JUnit test for storage and task list

Lanreath committed:
Add JavaDoc comments

Lanreath committed:
Change to follow conding standards

Lanreath committed:
Added find command

Lanreath committed:
Merge branch 'branch-A-CodingStandard'

Lanreath committed:
Fix naming errors

Lanreath committed:
Merge branch 'branch-Level-9'

Varstak commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,102 @@
+     * @param filePath the path of the file to save and load tasks from.
+     * @throws DukeException
+     */
+    public Storage(String filePath) throws DukeException{
```
Is there a missing whitespace between `DukeException` and `{`

Lanreath replied:
Good catch!

Varstak commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,116 @@
+     * @return the LocalDate object.
+     * @throws DukeException 
+     */
+    public static LocalDate parseDate(String date) throws DukeException {
```
Would the method name `stringToLocalDate` instead of `parseDate` be more clear?

Lanreath replied:
Using an action verb for a method name seems more suitable.

Varstak commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,116 @@
+                Integer i = Integer.parseInt(split[1]);
+                return new Delete(i);
+            } catch (NumberFormatException e) {
+                throw new DukeException("Please specify a valid task number.");
+            }
+        } else if (split[0].equals("find")) {
+            if (split.length == 1) {
+                throw new DukeException("Please specify the keyword you want to find.");
+            }
+            return new Find(split[1]);
+        } else if (split[0].equals("todo")) {
+            if (split.length == 1) {
+                throw new DukeException("The description of a todo cannot be empty.");
+            }
+            return new CreateTodo(split[1]);
+        } else if (split[0].equals("deadline")) {
+            if (split.length == 1) {
+                throw new DukeException("The description of a deadline cannot be empty.");
+            }
+            String[] tokens = split[1].split(" /by ", 2);
+            if (tokens.length == 1) {
+                throw new DukeException("Please provide a deadline for this task.");
+            }
+            return new CreateDeadline(tokens[0], parseDate(tokens[1]));
+        } else if (split[0].equals("event")) {
+            if (split.length == 1) {
+                throw new DukeException("The description of an event cannot be empty.");
+            }
+            String[] tokens = split[1].split(" /from ", 2);
+            if (tokens.length == 1) {
+                throw new DukeException("Please provide a start time for this event.");
+            }
+            String[] tokens2 = tokens[1].split(" /to ", 2);
+            if (tokens2.length == 1) {
+                throw new DukeException("Please provide an end time for this event.");
+            }
+            return new CreateEvent(tokens[0], parseDate(tokens2[0]), parseDate(tokens2[1]));
+        } else {
+            throw new DukeException("I'm sorry, but I don't know what that means :-(");
+        }
```
Could using `switch` statements help with additions to features in the future?

Lanreath replied:
That is a great suggestion, will include it in the next version 😄

Varstak commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,140 @@
+    /**
+     * Display line to separate user input and output.
+     */
+    public void showLine() {
+        System.out.println("----------------------------------------");
+    }
+
+    /**
+     * Display goodbye message.
+     */
+    public void showBye() {
+        System.out.println("Bye. Hope to see you again soon!");
+    }
```
Any reason you implemented these methods rather than saving the strings as constants?

Varstak reviewed (commented):
Not too many coding standard issues. Keep it up !!! :smile::smile::smile:

Varstak commented on src/test/java/duke/StorageTest.java:
```diff
@@ -0,0 +1,72 @@
+        } catch (IOException e) {
+            fail("IOException thrown");
+        } catch (DukeException e) {
+            fail("DukeException thrown");
```
Could making this string a constant in the parent `DukeException` class make the code easier to read and reduce bugs? I noticed you repeated the same string multiple times in this class.

Lanreath replied:
I wanted to leave room for future custom messages, but that could work if the messages end up staying the same.

Lanreath committed:
Change class member comments

immanuelhume commented:
Hi I came across your project and found it really cool since it is written in Java, it is 100% memory safe and blazingly fast :rocket: reliable software. so I would like to license DukePro to use it for my new startup, we want to switch from notion written in slow JavaScript to DukePro, can you please contact me and we can work out an arrangement thank you, regards.

Lanreath committed:
Add GUI and refactor text UI

Lanreath committed:
Redeclared missing ui object

Lanreath committed:
Added assertions for task creation and deletion.

Tasks being stored correctly is essential to the user retrieving information.
Using assertions prevents issues from occuring runtime.

Lanreath committed:
Refactor parser for better code quality

Readability for command parser was low with varying levels of abstraction.
Repeated code has been consolidated and variables renamed.

Lanreath committed:
Replace list iteration with stream methods

Lanreath committed:
Merge pull request #1 from Lanreath/branch-A-Assertions

Added assertions for task creation and deletion.

Lanreath committed:
Merge branch 'master' into branch-A-CodeQuality

Lanreath committed:
Merge pull request #2 from Lanreath/branch-A-CodeQuality

Refactor parser for better code quality

Lanreath committed:
Merge branch 'master' into branch-A-Streams

Lanreath committed:
Merge pull request #3 from Lanreath/branch-A-Streams

Replace list iteration with stream methods

Lanreath committed:
Added yml file for continuous integration

Lanreath committed:
Stage missing yml file for commit

Lanreath committed:
Fix wrong directory in test file

Lanreath committed:
Test IO exception handling in CI

Lanreath committed:
Remove folder and add file deletion at end of tests

Lanreath committed:
Fix error in file directory

Lanreath committed:
Remove IO redirection tests

Lanreath committed:
Merge pull request #4 from Lanreath/branch-A-CI

Added yml file for continuous integration

Lanreath committed:
Add simplified command syntax and user aliases

Improve speed of accessing commands by shortening them to single letters.
User customisation of commands
Future support for saving of aliases and args in any order

Lanreath committed:
Remove unnecessary imports

Lanreath committed:
Fix bye user input not exiting gui

Check command return value and concurrently close application
on one second delay.

Lanreath committed:
Rewrite user output, customise fonts, and reshape avatar images

Lanreath committed:
Merge branch 'branch-A-Personality'

Lanreath committed:
Declare 4 types of duke exceptions to handle errors

Lanreath committed:
Add tests for parser and ui

Lanreath committed:
Fix incorrect merge from branch-A-Personality

Lanreath committed:
Add deleted stage show call

Lanreath committed:
Fix expected test results and gui exit on bye command

Lanreath committed:
Rename product and minor change to error message

Lanreath committed:
Add user guide and ui screenshot

Lanreath committed:
Change expected test results

Lanreath committed:
Change docs to imporve code quality

Lanreath committed:
Change output for empty task list
